Clean-up and document symex_function_call [DOC-150]

romainbrenguier opened the pull request:
Make a couple of functions static and add documentation where missing.
 
<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [na] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [na] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [na] My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- [na] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

romainbrenguier committed:
Make pop_frame static

This function is only used within the same cpp file.
Making it static simplifies the global interface and is potentially more efficient.

romainbrenguier committed:
Make locality a static function

locality is only used in symex_function_call.cpp, it can be made static.
This simplifies the symex interface and is potentially more efficient.

romainbrenguier committed:
Document symex_function_call functions

This documents (mostly in goto_symex.h) functions that are defined in symex_function_call.cpp.

tautschnig reviewed (changes requested):
Could "Move and make static symex_end_of_function" please be skipped? If we go this route, we will eventually have a single file containing all source code.

romainbrenguier commented:
@tautschnig 
> If we go this route, we will eventually have a single file containing all source code.

That would be better because more efficient. 
More seriously, I think the better way to avoid having all the code in one file would be to split goto_symex in several smaller meaningful classes with their own header. I find having several cpp file for the same header confusing, and I'm not sure of the advantage.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 13ce1dea3491d4c3a37f1de8eb67b56890c26a6c).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/99785216

tautschnig commented:
> More seriously, I think the better way to avoid having all the code in one file would be to split goto_symex in several smaller meaningful classes with their own header. I find having several cpp file for the same header confusing, and I'm not sure of the advantage.

Fully agreed, just the other day I had to make some changes in `src/solvers/strings/`... But then at least the existing cpp files are a starting point and really just the classes should be added in.

romainbrenguier commented:
@tautschnig 
> But then at least the existing cpp files are a starting point and really just the classes should be added in.

Yes, so I don't see the problem with that particular commit, since it's reducing the dependencies between different cpp files, it will make it a little bit easier.

tautschnig commented:
> I don't see the problem with that particular commit

It's moving code that is entirely specific to dealing with function calls to `symex_main` from `symex_function_call`.

romainbrenguier commented:
> It's moving code that is entirely specific to dealing with function calls to `symex_main` from `symex_function_call`.

OK, I see the point and will drop it. I still think grouping `function_call` and `end_function` is artificial since they don't depend on each other, but refactoring `goto_symext` can wait a bit.

tautschnig commented:
> I still think grouping `function_call` and `end_function` is artificial since they don't depend on each other, [...]

This may be true as far as compiling code is concerned, but at a more abstract level `end_function` needs to be consistent with what `function_call` does, e.g., in terms of setting up frames. It's actually bad that there is no dependency, because it means it's all down to humans checking consistency.

tautschnig commented on src/goto-symex/goto_symex.h:
```diff
@@ -271,25 +271,41 @@ class goto_symext
   virtual void loop_bound_exceeded(statet &, const exprt &guard);
 
-  // function calls
+  /// Assuming \p state is currently pointing on a return instruction, assign
```
The state doesn't point, maybe the program counter inside it?

tautschnig commented on src/goto-symex/goto_symex.h:
```diff
@@ -271,25 +271,41 @@ class goto_symext
-  // function calls
+  /// Assuming \p state is currently pointing on a return instruction, assign
+  /// the value in that return to the top frame's \p return_value field.
   void return_assignment(statet &);
```
Is this actually dead code? return-value removal should render it such.

romainbrenguier replied:
Indeed it looks like it, I've added a commit to replace the call to this by an UNREACHABLE

tautschnig commented on src/goto-symex/goto_symex.h:
```diff
@@ -300,6 +316,12 @@ class goto_symext
     unsigned thread_nr,
     unsigned unwind);
 
+  /// Iterates over \p arguments and assign them to the parameters, which are
```
assigns

tautschnig commented on src/goto-symex/goto_symex.h:
```diff
@@ -300,6 +316,12 @@ class goto_symext
     unsigned unwind);
 
+  /// Iterates over \p arguments and assign them to the parameters, which are
+  /// symbols whose name and type are deduce from the type of \p goto_function.
```
deduced

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: f7d8891f7e530ad35c69896068387c02b5f79c7f).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/99811235

romainbrenguier committed:
Remove unused case of return assignment in symex

The return instructions should have been removed by return-value
removal, so they should not reach symex.
We replace this case by an unreachable statement and remove the method
that is now unused.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 16c816516fdbe8f0e31808078980c73dfe754f59).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/99899024

romainbrenguier commented:
@peterschrammel this is ready for review

peterschrammel commented on src/goto-symex/goto_symex.h:
```diff
@@ -271,27 +271,37 @@ class goto_symext
 
+  /// Symbolic execution of a function call.
+  /// Only functions that are symbols are supported, see
+  /// \link goto_symext::symex_function_call_symbol \endlink
```
\ref doesn't work here?

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 7dc47a4c6681ea61b562e3ad7edb96a3f55e5034).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/100121863
